File: src-tauri/src/models/recurring.rs
// File: src-tauri/src/models/recurring.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct RecurringTransaction {
    pub id: i64,
    pub name: String,
    pub description: Option<String>,
    pub transaction_type: String, // INCOME, EXPENSE, TRANSFER
    pub amount: f64,
    pub account_id: i64,
    pub to_account_id: Option<i64>,
    pub category_id: Option<i64>,
    pub frequency: String,  // DAILY, WEEKLY, MONTHLY, YEARLY, CUSTOM
    pub interval_days: i64, // For CUSTOM frequency
    pub start_date: String,
    pub end_date: Option<String>,
    pub next_execution_date: String,
    pub is_active: bool,
    pub last_executed_date: Option<String>,
    pub execution_count: i64,
    pub created_at: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateRecurringTransactionInput {
    pub name: String,
    pub description: Option<String>,
    pub transaction_type: String,
    pub amount: f64,
    pub account_id: i64,
    pub to_account_id: Option<i64>,
    pub category_id: Option<i64>,
    pub frequency: String,
    pub interval_days: Option<i64>, // Only for CUSTOM frequency
    pub start_date: String,
    pub end_date: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateRecurringTransactionInput {
    pub id: i64,
    pub name: Option<String>,
    pub description: Option<String>,
    pub amount: Option<f64>,
    pub frequency: Option<String>,
    pub interval_days: Option<i64>,
    pub end_date: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct RecurringTransactionWithDetails {
    #[serde(flatten)]
    pub recurring: RecurringTransaction,
    pub account_name: String,
    pub to_account_name: Option<String>,
    pub category_name: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct UpcomingExecution {
    pub recurring_id: i64,
    pub name: String,
    pub transaction_type: String,
    pub amount: f64,
    pub next_execution_date: String,
    pub days_until_execution: i64,
}

File: src/types/recurring.ts
// File: src/types/recurring.ts
export type RecurringFrequency =
  | "DAILY"
  | "WEEKLY"
  | "MONTHLY"
  | "YEARLY"
  | "CUSTOM";

export type AmountMode = "FIXED" | "VARIABLE";

export interface RecurringTransaction {
  id: number;
  name: string;
  description: string | null;
  transaction_type: "INCOME" | "EXPENSE" | "TRANSFER";
  amount: number;
  account_id: number;
  to_account_id: number | null;
  category_id: number | null;
  frequency: RecurringFrequency;
  interval_days: number;
  start_date: string;
  end_date: string | null;
  next_execution_date: string;
  is_active: boolean;
  last_executed_date: string | null;
  execution_count: number;
  created_at: string;
  // V1.2.0
  amount_mode: AmountMode;
  resume_date: string | null;
  active_months: string | null; // "1,2,3,10,11,12"
  auto_approve: boolean;
}

export interface CreateRecurringTransactionInput {
  name: string;
  description: string | null;
  transaction_type: "INCOME" | "EXPENSE" | "TRANSFER";
  amount: number;
  account_id: number;
  to_account_id: number | null;
  category_id: number | null;
  frequency: RecurringFrequency;
  interval_days: number | null;
  start_date: string;
  end_date: string | null;
  // V1.2.0
  amount_mode?: AmountMode;
  active_months?: string | null;
  auto_approve?: boolean;
}

export interface RecurringExecutionLog {
  id: number;
  recurring_id: number;
  execution_date: string;
  status: "SUCCESS" | "SKIPPED" | "FAILED" | "VARIABLE_PENDING";
  amount: number | null;
  transaction_id: number | null;
  notes: string | null;
  created_at: string;
}
